### PR TITLE
Feature handle left players correctly

### LINKED ISSUE #193: Handle it correctly if a player leaves during game
If a player leaves during the game, the other players should be able to continue playing.

### PR DESCRIPTION
**Pull request by BacLuc**
closes #193 
Issue: https://github.com/sopra-fs22-group-36/screw-your-neighbor-server/issues/160
Issue: https://github.com/sopra-fs22-group-36/screw-your-neighbor-server/issues/169

### CODE REVIEW

**Comment by salwil** [comment]
Impressive work, thanks a lot!

File: src/main/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/sideeffects/ParticipationEventHandler.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  @HandleBeforeSave
+  public void onBeforeSave(Participation participation) {
+    Participation previousStateOf =
+        oldStateFetcher.getPreviousStateOf(Participation.class, participation.getId());
+    if (previousStateOf.isActive() == participation.isActive()) {
+      return;
+    }
+    if (participation.isActive()) {
+      return;
```
[the remark this answers is not in the document]
Comment (BacLuc) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/sideeffects/ParticipationEventHandler.java:37]: ITS the Default of the flag

File: src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/api/RoundApiTest.java [hunk 2]
```diff
@@ -113,4 +115,73 @@ void return_correct_isStacked() {
...
+            .finishRound()
+            .finishMatch()
+            .build();
+
+    Participation participation3 =
+        game.getParticipations().stream()
+            .filter(participation -> PLAYER_NAME_3.equals(participation.getPlayer().getName()))
+            .findFirst()
+            .orElseThrow();
+    participation3.setActive(false);
```
[the remark this answers is not in the document]
Comment (BacLuc) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/api/RoundApiTest.java:167]: If clicks leave, the Partizipation ist Patches top inactive

File: src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/integration/CardEventHandlerTest.java [hunk 3]
```diff
@@ -358,6 +466,78 @@ void play_last_card_of_battling_round_with_round_stacked() {
...
+
+    cardEventHandler.handleAfterSave(lastCardPlayer1);
+
+    assertThat(game.getLastMatch().map(Match::getMatchNumber), is(of(2)));
+    assertThat(match.getLastRound().map(Round::getRoundNumber), is(of(4)));
+    assertThat(game.getMatches(), hasSize(2));
+
+    assertThat(handPlayer2.getNumberOfWonTricks(), is(0));
+    assertThat(handPlayer1.getNumberOfWonTricks(), is(3));
+    assertThat(handPlayer3.getNumberOfWonTricks(), is(1));
```
[the remark this answers is not in the document]
Comment (BacLuc) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/integration/CardEventHandlerTest.java:538]: Feel free to implement that

File: src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/api/RoundApiTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (salwil) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/api/RoundApiTest.java:167]: Ah right, I was always thinking about the closing of the browser window. That we can't handle, I assume.

File: src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/integration/CardEventHandlerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (salwil) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/integration/CardEventHandlerTest.java:538]: I'd implement it such that the tricks go to players with name lucius and salome.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/validation/ParticipationValidator.java [hunk 4]
```diff
@@ -11,24 +14,46 @@
...
+    if (game.getParticipations().stream().filter(Participation::isActive).count()
+        > MAX_NUMBER_OF_PLAYERS) {
       throw new HttpClientErrorException(
           HttpStatus.UNPROCESSABLE_ENTITY,
           "Not more then %s players per game are allowed".formatted(MAX_NUMBER_OF_PLAYERS));
     }
   }
+
+  @HandleBeforeSave
+  public void onBeforeSaveParticipation(Participation participation) {
```
[the remark this answers is not in the document]
Comment (BacLuc) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/validation/ParticipationValidator.java:49]: No, one is @HandleBeforeSave (for PATCH), one is @HandleBeforeCreate

File: src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/integration/CardEventHandlerTest.java [hunk 5]
```diff
@@ -234,6 +280,68 @@ void puts_game_to_state_closed_for_last_card_in_last_match() {
...
+
+    assertThat(game.getMatches(), hasSize(9));
+
+    cardEventHandler.handleAfterSave(notYetPlayedCard);
+
+    Game updatedGame = gameRepository.findById(game.getId()).orElseThrow();
+    assertThat(updatedGame.getMatches(), hasSize(9));
+    assertThat(updatedGame.getGameState(), is(CLOSED));
+  }
+
```
[the remark this answers is not in the document]
Comment (BacLuc) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/integration/CardEventHandlerTest.java:344]: If a player left the game, not every player has played a card in this round.
But we still have to trigger all the logic (create new round, new match or close the game).

File: src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/sideeffects/ParticipationEventHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    previousParticipation.setActive(true);
+    participation.setActive(false);
+
+    participationEventHandler.onBeforeSave(participation);
+
+    verifyNoInteractions(handEventHandler, cardEventHandler);
+  }
+
+  @Test
+  void calls_handEventHandler_but_not_cardEventHandler_if_no_round() {
```
[the remark this answers is not in the document]
Comment (BacLuc) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/sideeffects/ParticipationEventHandlerTest.java:146]: we have to do the same when there is no round at all and when there is a round without played card.
You can see that with the assertions in the tests

**Pull request merged**